MediaBrowser.Providers/Manager/MetadataService.cs

```diff
@@ -543,7 +542,7 @@ private ItemUpdateType UpdateOfficialRating(TItemType item, IList<BaseItem> chil
         protected IEnumerable<IMetadataProvider> GetProviders(BaseItem item, LibraryOptions libraryOptions, MetadataRefreshOptions options, bool isFirstRefresh, bool requiresRefresh)
         {
             // Get providers to refresh
-            var providers = ((ProviderManager)ProviderManager).GetMetadataProviders<TItemType>(item, libraryOptions).ToList();
+            var providers = ((ProviderManager)ProviderManager).GetMetadataProviders<TItemType>(item, libraryOptions);
```

Comment:
`providers` is enumerated 4 times in this function and then again at least 3 times in the caller.

Reply:
Sorry, I can only count 2 enumerations of `providers`, both with the `Where` clause. What am I missing? Also this was also done before, however it used `ToList` causing an allocation. 

I wrote the following benchmark to show the effects of my changes (see below). The results:
|      Method |     Mean |    Error |   StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------:|---------:|---------:|--------:|------:|------:|----------:|
|      ToList | 26.29 us | 0.517 us | 0.690 us | 13.0310 |     - |     - |  39.99 KB |
| IEnumerable | 28.61 us | 0.547 us | 0.651 us | 11.6272 |     - |     - |  35.67 KB |

So that's a 2.32 microsecond (8.82%) increase in CPU for a 4.32KB (10.8%) reduction in allocations.

While there may be 1 or 2 tweaks I can make to get the CPU and allocations down I think the changes as is are still an improvement (we can debate) but more importantly my changes don't loop over any one collection more then before (again, assuming I'm not missing something).

```cs
class Program
{
    static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<AllocationTest>();
        Console.WriteLine(summary);
    }
}

[MemoryDiagnoser]
public class AllocationTest
{
    const int Max = 1000;

    readonly List<int> data;
    
    public AllocationTest()
    {
        data = new List<int>(Max);
        for (int i = 0; i < Max; i++)
        {
            data.Add(i);
        }
    }

    [Benchmark]
    public void ToList()
    {
        List<int> numbers = data;
        var evenNumbers = data.Where(i => i % 2 == 0).ToList();

        if (evenNumbers.Count == 0)
        {
            numbers = new List<int>();
        }
        else
        {
            var doubles = evenNumbers.OfType<double>().Any();

            var floats = evenNumbers.OfType<float>().Any();

            var chars = evenNumbers.OfType<char>().Any();

            if (doubles || floats || chars)
            {
                throw new Exception("What?");
            }

            numbers = numbers.Where(i => i < 10).ToList();
        }
    }

    [Benchmark]
    public void IEnumerable()
    {
        IEnumerable<int> numbers = data;
        var evenNumbers = data.Where(i => i % 2 == 0);

        if (!evenNumbers.Any())
        {
            numbers = new List<int>();
        }
        else
        {
            var doubles = evenNumbers.OfType<double>().Any();

            var floats = evenNumbers.OfType<float>().Any();

            var chars = evenNumbers.OfType<char>().Any();

            if (doubles || floats || chars)
            {
                throw new Exception("What?");
            }

            numbers = numbers.Where(i => i < 10);
        }
    }
}
```

Reply:
Actually I was able to make a small tweak and get both the CPU and allocations down. I'll update my commit. Still, not seeing how I increased enumerations:
|       Method |     Mean |    Error |   StdDev |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |---------:|---------:|---------:|--------:|------:|------:|----------:|
|       ToList | 25.99 us | 0.146 us | 0.130 us | 13.0310 |     - |     - |   40952 B |
|  IEnumerable | 28.81 us | 0.412 us | 0.321 us | 11.6272 |     - |     - |   36528 B |
| IEnumerable2 | 15.56 us | 0.158 us | 0.148 us |  0.0916 |     - |     - |     288 B |

```cs
[Benchmark]
public void IEnumerable2()
{
    IEnumerable<int> numbers = data;
    var evenNumbers = data.Where(i => i % 2 == 0);

    if (!evenNumbers.Any())
    {
        numbers = new List<int>();
    }
    else
    {
        var doubles = evenNumbers.Any(i => i is double);

        var floatOrChar = evenNumbers.Any(i => i is float || i is char);

        if (doubles || floatOrChar)
        {
            throw new Exception("What?");
        }

        numbers = numbers.Where(i => i < 10);
    }
}
```

Reply:
Fixed.